martin/src/bin/martin-cp.rs

```diff
@@ -566,12 +600,102 @@ async fn main() {
 
 #[cfg(test)]
 mod tests {
+    use super::*;
+    use async_trait::async_trait;
+    use insta::assert_yaml_snapshot;
+    use martin::TileSources;
+    use martin_core::tiles::{MartinCoreResult, Source, UrlQuery};
+    use martin_tile_utils::{Encoding, Format};
+    use rstest::{fixture, rstest};
     use std::str::FromStr;
+    use tilejson::{TileJSON, tilejson};
 
-    use insta::assert_yaml_snapshot;
-    use rstest::rstest;
+    #[derive(Debug, Clone)]
+    pub struct MockSource {
+        pub id: &'static str,
+        pub tj: TileJSON,
+        pub data: TileData,
+    }
 
-    use super::*;
+    #[async_trait]
+    impl Source for MockSource {
+        fn get_id(&self) -> &str {
+            self.id
+        }
+
+        fn get_tilejson(&self) -> &TileJSON {
+            &self.tj
+        }
+
+        fn get_tile_info(&self) -> TileInfo {
+            TileInfo::new(Format::Mvt, Encoding::Uncompressed)
+        }
+
+        fn clone_source(&self) -> BoxedSource {
+            Box::new(self.clone())
+        }
+
+        async fn get_tile(
+            &self,
+            _xyz: TileCoord,
+            _url_query: Option<&UrlQuery>,
+        ) -> MartinCoreResult<TileData> {
+            Ok(self.data.clone())
+        }
+    }
+
+    #[fixture]
+    fn many_sources() -> TileSources {
+        TileSources::new(vec![vec![
+            Box::new(MockSource {
+                id: "test_source",
+                tj: tilejson! { tiles: vec![], bounds: Bounds::from_str("-110.0,20.0,-120.0,80.0").unwrap() },
+                data: Vec::default(),
+            }),
+            Box::new(MockSource {
+                id: "test_source2",
+                tj: tilejson! { tiles: vec![], bounds: Bounds::from_str("-130.0,40.0,-170.0,10.0").unwrap() },
+                data: Vec::default(),
+            }),
+            Box::new(MockSource {
+                id: "unrequested_source",
+                tj: tilejson! { tiles: vec![], bounds: Bounds::from_str("-150.0,40.0,-120.0,10.0").unwrap() },
+                data: Vec::default(),
+            }),
+        ]])
+    }
+
+    #[fixture]
+    fn one_source() -> TileSources {
+        TileSources::new(vec![vec![Box::new(MockSource {
+            id: "test_source",
+            tj: tilejson! { tiles: vec![], bounds: Bounds::from_str("-120.0,30.0,-110.0,40.0").unwrap() },
+            data: Vec::default(),
+        })]])
+    }
+
+    #[fixture]
+    fn source_wo_bounds() -> TileSources {
+        TileSources::new(vec![vec![Box::new(MockSource {
+            id: "test_source",
+            tj: tilejson! { tiles: vec![] },
+            data: Vec::default(),
+        })]])
+    }
+
+    #[rstest]
+    #[case::one_source(one_source(), "test_source", vec![Bounds::from_str("-120.0,30.0,-110.0,40.0").unwrap()])]
+    #[case::many_sources(many_sources(), "test_source,test_source2", vec![Bounds::from_str("-110.0,20.0,-120.0,80.0").unwrap(), Bounds::from_str("-130.0,40.0,-170.0,10.0").unwrap()])]
```

Comment:
lets also test the case for multi source with one being without a bound

```suggestion
    #[case::many_sources(many_sources(), "test_source,test_source2", vec![Bounds::from_str("-110.0,20.0,-120.0,80.0").unwrap(), Bounds::from_str("-130.0,40.0,-170.0,10.0").unwrap()])]
    #[case::many_sources(many_sources(), "unbounded_source", vec![Bounds::MAX_TILED])]
    #[case::many_sources(many_sources(), "test_source,unbounded_source", vec![Bounds::from_str("-110.0,20.0,-120.0,80.0").unwrap(), Bounds::MAX_TILED])]
```

Reply:
Will hold off updating tests until the approach is decided above, but I was curious here if ordering matters? Looks like that was the intention with the `_rev` tests.

Reply:
ordering should not matter in downstream code, but this makes us check this if the code changes.